File: business/src/main/java/com/pij/lottieshow/detail/LottieViewModel.java
package com.pij.lottieshow.detail;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.pij.lottieshow.interactor.LottieSink;
import com.pij.lottieshow.interactor.Serializer;
import com.pij.lottieshow.model.LottieFile;

import rx.Observable;
import rx.subjects.PublishSubject;

import static rx.Observable.empty;
import static rx.Observable.merge;

class LottieViewModel {

    private final PublishSubject<LottieFile> lottieToLoad = PublishSubject.create();
    private final PublishSubject<LottieFile> lottieToAdd = PublishSubject.create();
    private final PublishSubject<Throwable> errors = PublishSubject.create();
    private final Observable<LottieFile> lottieToShow;
    private final Observable<String> animationToShow;

    @SuppressWarnings("WeakerAccess")
    public LottieViewModel(@NonNull LottieSink sink, @NonNull Serializer serializer) {

        Observable<LottieFile> addedLottie = lottieToAdd.flatMap(file -> Observable.just(file)
                                                                                   .doOnNext(sink::add)
                                                                                   .doOnError(errors::onNext)
                                                                                   .onErrorResumeNext(empty()));
        //noinspection UnnecessaryLocalVariable
        Observable<LottieFile> loadedLottie = lottieToLoad;
        lottieToShow = merge(loadedLottie, addedLottie);

        animationToShow = lottieToShow.map(file -> file == null ? null : file.id())
                                      .concatMap(input -> serializer.open(input)
                                                                    .toObservable()
                                                                    .doOnError(errors::onNext)
                                                                    .onErrorResumeNext(empty()));
    }

    @SuppressWarnings("WeakerAccess")
    public void loadLottie(@Nullable LottieFile newFile) {
        lottieToLoad.onNext(newFile);
    }

    @SuppressWarnings("WeakerAccess")
    public Observable<String> showAnimation() {
        return animationToShow;
    }

    @SuppressWarnings("WeakerAccess")
    public Observable<Throwable> showLoadingError() {
        return errors;
    }

    @SuppressWarnings("WeakerAccess")
    public void addLottie(@NonNull LottieFile newFile) {
        lottieToAdd.onNext(newFile);
    }

    @SuppressWarnings("WeakerAccess")
    public Observable<LottieFile> showLottie() {
        return lottieToShow;
    }

}

File: business/src/main/java/com/pij/lottieshow/detail/LottieViewModel.kt
package com.pij.lottieshow.detail

import com.pij.lottieshow.interactor.LottieSink
import com.pij.lottieshow.interactor.Serializer
import com.pij.lottieshow.model.LottieFile
import rx.Observable
import rx.Observable.empty
import rx.Observable.merge
import rx.subjects.PublishSubject
import java.net.URI

class LottieViewModel(private val sink: LottieSink, private val serializer: Serializer) {

    private val lottieToLoad = PublishSubject.create<LottieFile>()
    private val lottieToAdd = PublishSubject.create<LottieFile>()
    private val errors = PublishSubject.create<Throwable>()
    private val lottieToShow: Observable<LottieFile>
    private val animationToShow: Observable<String>

    init {

        val addedLottie = lottieToAdd.flatMap {
            Observable.just(it)
                    .doOnNext { sink.add(it) }
                    .doOnError { errors.onNext(it) }
                    .onErrorResumeNext(empty<LottieFile>())
        }

        val loadedLottie = lottieToLoad
        lottieToShow = merge(loadedLottie, addedLottie)

        animationToShow = lottieToShow.map<URI> { it?.id() }
                .concatMap {
                    serializer.open(it)
                            .toObservable()
                            .doOnError { errors.onNext(it) }
                            .onErrorResumeNext(empty<String>())
                }
    }

    fun loadLottie(newFile: LottieFile?) {
        lottieToLoad.onNext(newFile)
    }

    fun showAnimation(): Observable<String> {
        return animationToShow
    }

    fun showLoadingError(): Observable<Throwable> {
        return errors
    }

    fun addLottie(newFile: LottieFile) {
        lottieToAdd.onNext(newFile)
    }

    fun showLottie(): Observable<LottieFile> {
        return lottieToShow
    }

}
